drivers/vulkan/rendering_context_driver_vulkan.cpp

```diff
@@ -40,21 +40,340 @@
+uint64_t RenderingContextDriverVulkan::get_driver_total_memory() const {
+	return driver_memory_total_memory.get();
+}
+
+uint64_t RenderingContextDriverVulkan::get_driver_allocation_count() const {
+	return driver_memory_total_alloc_count.get();
+}
+
+uint64_t RenderingContextDriverVulkan::get_driver_memory_by_object_type(uint32_t p_type) const {
+	uint64_t ret = 0;
+	for (uint32_t i = 0; i < VK_TRACKED_SYSTEM_ALLOCATION_SCOPE_COUNT; i++) {
+		ret += driver_memory_tracker[p_type][i].get();
+	}
+
+	return ret;
+}
+
+uint64_t RenderingContextDriverVulkan::get_driver_allocs_by_object_type(uint32_t p_type) const {
+	uint64_t ret = 0;
+	for (uint32_t i = 0; i < VK_TRACKED_SYSTEM_ALLOCATION_SCOPE_COUNT; i++) {
+		ret += driver_memory_allocation_count[p_type][i].get();
+	}
+
+	return ret;
+}
+#endif
+
+#if defined(VK_TRACK_DEVICE_MEMORY)
+void RenderingContextDriverVulkan::memory_report_callback(const VkDeviceMemoryReportCallbackDataEXT *p_callback_data, void *p_user_data) {
+	if (!p_callback_data) {
+		return;
+	}
+	const RenderingContextDriverVulkan::VkTrackedObjectType obj_type = vk_object_to_tracked_object(p_callback_data->objectType);
+	uint64_t obj_id = p_callback_data->memoryObjectId;
+
+	if (p_callback_data->type == VK_DEVICE_MEMORY_REPORT_EVENT_TYPE_ALLOCATE_EXT) {
+		// Realloc, update size
+		if (memory_report_table.has(obj_id)) {
+			memory_report_total_memory.sub(memory_report_table[obj_id]);
+			memory_report_mem_usage[obj_type].sub(memory_report_table[obj_id]);
+
+			memory_report_total_memory.add(p_callback_data->size);
+			memory_report_mem_usage[obj_type].add(p_callback_data->size);
+
+			memory_report_table[p_callback_data->memoryObjectId] = p_callback_data->size;
+		} else {
+			memory_report_table[obj_id] = p_callback_data->size;
+
+			memory_report_total_alloc_count.increment();
+			memory_report_allocation_count[obj_type].increment();
+			memory_report_mem_usage[obj_type].add(p_callback_data->size);
+			memory_report_total_memory.add(p_callback_data->size);
+		}
+	} else if (p_callback_data->type == VK_DEVICE_MEMORY_REPORT_EVENT_TYPE_FREE_EXT) {
+		if (memory_report_table.has(obj_id)) {
+			memory_report_total_alloc_count.decrement();
+			memory_report_allocation_count[obj_type].decrement();
+			memory_report_mem_usage[obj_type].sub(p_callback_data->size);
+			memory_report_total_memory.sub(p_callback_data->size);
+
+			memory_report_table.remove(memory_report_table.find(obj_id));
+		}
+	}
+}
+#endif
+
+VkAllocationCallbacks *RenderingContextDriverVulkan::get_allocation_callbacks(VkObjectType p_type) {
+#if !defined(VK_TRACK_DRIVER_MEMORY)
+	return nullptr;
+#else
+	struct TrackedMemHeader {
+		size_t size;
+		VkSystemAllocationScope allocation_scope;
+		VkTrackedObjectType type;
+	};
+	VkAllocationCallbacks tracking_callbacks = {
+		// Allocation function
+		nullptr,
+		[](
+				void *p_user_data,
+				size_t size,
+				size_t alignment,
+				VkSystemAllocationScope allocation_scope) -> void * {
```

Comment:
This lambda seems to not compile properly on arm32.

Here's a log excerpt from a failed Android editor arm32 build:
```
/root/sdk/ndk/23.2.8568313/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ -o editor/editor_string_names.android.editor.arm32.o -c -std=gnu++17 -fno-exceptions -target armv7a-linux-androideabi21 -fpic -ffunction-sections -funwind-tables -fstack-protector-strong -fvisibility=hidden -fno-strict-aliasing -march=armv7-a -mfloat-abi=softfp -ffp-contract=off -O2 -w -DTOOLS_ENABLED -DDEBUG_ENABLED -DNDEBUG -DENGINE_UPDATE_CHECK_ENABLED -DNO_EDITOR_SPLASH -D__ARM_ARCH_7__ -D__ARM_ARCH_7A__ -D__ARM_NEON__ -DANDROID_ENABLED -DUNIX_ENABLED -DVULKAN_ENABLED -DRD_ENABLED -DGLES3_ENABLED -DMINIZIP_ENABLED -DBROTLI_ENABLED -DTHREADS_ENABLED -DCLIPPER2_ENABLED -DZSTD_STATIC_LINKING_ONLY -DUSE_VOLK -DVK_USE_PLATFORM_ANDROID_KHR -DHAVE_MNTENT -Ithirdparty/freetype/include -Ithirdparty/libpng -Ithirdparty/volk -Ithirdparty/vulkan -Ithirdparty/vulkan/include -Ithirdparty/zstd -Ithirdparty/zlib -Ithirdparty/clipper2/include -Ithirdparty/brotli/include -Iplatform/android -I. editor/editor_string_names.cpp
[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:240:3: [0m[0;1;31merror: [0m[1mno viable conversion from '(lambda at drivers/vulkan/rendering_context_driver_vulkan.cpp:240:3)' to 'PFN_vkAllocationFunction' (aka 'void *(*)(void *, unsigned int, unsigned int, VkSystemAllocationScope) __attribute__((pcs("aapcs-vfp")))')[0m
                [](
[0;1;32m                ^~~
[0m[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:240:3: [0m[0;1;30mnote: [0mcandidate function[0m
[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:272:3: [0m[0;1;31merror: [0m[1mno viable conversion from '(lambda at drivers/vulkan/rendering_context_driver_vulkan.cpp:272:3)' to 'PFN_vkReallocationFunction' (aka 'void *(*)(void *, void *, unsigned int, unsigned int, VkSystemAllocationScope) __attribute__((pcs("aapcs-vfp")))')[0m
                [](
[0;1;32m                ^~~
[0m[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:272:3: [0m[0;1;30mnote: [0mcandidate function[0m
[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:306:3: [0m[0;1;31merror: [0m[1mno viable conversion from '(lambda at drivers/vulkan/rendering_context_driver_vulkan.cpp:306:3)' to 'PFN_vkFreeFunction' (aka 'void (*)(void *, void *) __attribute__((pcs("aapcs-vfp")))')[0m
                [](
[0;1;32m                ^~~
[0m[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:306:3: [0m[0;1;30mnote: [0mcandidate function[0m
[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:325:3: [0m[0;1;31merror: [0m[1mno viable conversion from '(lambda at drivers/vulkan/rendering_context_driver_vulkan.cpp:325:3)' to 'PFN_vkInternalAllocationNotification' (aka 'void (*)(void *, unsigned int, VkInternalAllocationType, VkSystemAllocationScope) __attribute__((pcs("aapcs-vfp")))')[0m
                [](
[0;1;32m                ^~~
[0m[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:325:3: [0m[0;1;30mnote: [0mcandidate function[0m
[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:331:3: [0m[0;1;31merror: [0m[1mno viable conversion from '(lambda at drivers/vulkan/rendering_context_driver_vulkan.cpp:331:3)' to 'PFN_vkInternalFreeNotification' (aka 'void (*)(void *, unsigned int, VkInternalAllocationType, VkSystemAllocationScope) __attribute__((pcs("aapcs-vfp")))')[0m
                [](
[0;1;32m                ^~~
[0m[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:331:3: [0m[0;1;30mnote: [0mcandidate function[0m
/root/sdk/ndk/23.2.8568313/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ -o editor/editor_help.android.editor.arm32.o -c -std=gnu++17 -fno-exceptions -target armv7a-linux-androideabi21 -fpic -ffunction-sections -funwind-tables -fstack-protector-strong -fvisibility=hidden -fno-strict-aliasing -march=armv7-a -mfloat-abi=softfp -ffp-contract=off -O2 -w -DTOOLS_ENABLED -DDEBUG_ENABLED -DNDEBUG -DENGINE_UPDATE_CHECK_ENABLED -DNO_EDITOR_SPLASH -D__ARM_ARCH_7__ -D__ARM_ARCH_7A__ -D__ARM_NEON__ -DANDROID_ENABLED -DUNIX_ENABLED -DVULKAN_ENABLED -DRD_ENABLED -DGLES3_ENABLED -DMINIZIP_ENABLED -DBROTLI_ENABLED -DTHREADS_ENABLED -DCLIPPER2_ENABLED -DZSTD_STATIC_LINKING_ONLY -DUSE_VOLK -DVK_USE_PLATFORM_ANDROID_KHR -DHAVE_MNTENT -Ithirdparty/freetype/include -Ithirdparty/libpng -Ithirdparty/volk -Ithirdparty/vulkan -Ithirdparty/vulkan/include -Ithirdparty/zstd -Ithirdparty/zlib -Ithirdparty/clipper2/include -Ithirdparty/brotli/include -Iplatform/android -I. editor/editor_help.cpp
5 errors generated.
```
(sorry for the ANSI codes mangled in)

Reply:
That's odd.

The signature for PFN_vkAllocationFunction is:

```cpp
typedef void* (VKAPI_PTR *PFN_vkAllocationFunction)(
    void*                                       pUserData,
    size_t                                      size,
    size_t                                      alignment,
    VkSystemAllocationScope                     allocationScope);
```

Which is defined in godot/thirdparty/vulkan/include/vulkan/vulkan_core.h

And our lambda is:

```cpp
		[](
				void *p_user_data,
				size_t size,
				size_t alignment,
				VkSystemAllocationScope allocation_scope) -> void * {
```

The only thing that comes to mind is that the definition of "size_t" may be different when parsing vulkan_core.h than when parsing the lambda in rendering_context_driver_vulkan.cpp

That's the most suspicious one since Vulkan has a tendency to define everything as uint64, while size_t should be 32-bit in arm32.

Is there an easy way I can repro this? (I have xp with compiling for Android, but I've been compiling for aarch64).

Does the problem go away if you change the size_t in the lambdas for uint64_t ? (and what about uint32_t?)